enhancements/insights/insights-operator-runtime.md

```diff
@@ -0,0 +1,532 @@
+The fingerprints that are executing in the container namespaces must be able to sanitize the extracted information and discard them if they deviate from the expected output.
+
+
+To make sure customer confidentiality is preserved, any new collected data is hashed before they are sent to Red Hat backends. Mechanisms are already in place so that access to the unobfuscated data in Red Hat is granted on a needs-to-know basis.
+
+#### Performance Risk & Mitigation
+
+##### Limited Restrictions for Fingerprints
+
+Since a fingerprint executable runs under a downgraded privilege model, with a limited process life-span, significantly more freedom is afforded to it, but some care is still necessary. Notably, a fingerprint executable can have a larger dependency set beyond the core language SDK to meet its needs. Additionally, fingerprint executables can leverage languages with a greater emphasis on productivity (e.g. Golang), provided the implementations still utilize input and output hardening techniques, and do not load potentially rogue code in the container. Each fingerprint implementation may even have a different language runtime than its peers. The following rules should be followed:
+
+1. __Inspecting code must be statically compiled and not rely on dynamic library loading.__ This is essentially already required to achieve stable execution on arbitrary container layers, including those with differing libc implementations, or even scratch containers with no additional libraries
+2. __All analysis I/O operations should be limit-bound.__ A container either through error or design may have files of arbitrary size, or files in a corrupt state. Any potentially unbounded operation (e.g. reading to the end of a file or following a reference data structure to its arbitrary end) is prohibited.
+3. __Memory and CPU usage should not be input derived.__ In addition to achieving system stability and performance goals, ensuring the overhead of a inspecting process does not increase significantly relative to input data is critical in preventing potential Denial-of-Service attacks relying on a specially crafted container filesystem.  
+
+##### No Impact on Workloads
+
+Inspecting containers must not disrupt the containers that are inspected and have an impact on their quality of service.
+The insights-operator-runtime is running with its own memory and CPU requirements and does not consume resources from the inspected container.
+
+##### Measurable Impact on Gathering Time & Payload
+
+The workload data is gathered every 12 hours by the Insights Operator.
+Gathering the runtime data will increase the overall executing time of the workload gatherer. The increase must remain reasonable.
+
+In the existing initial implementation of the insights-operator-runtime, it takes __~200 ms to scan a single container__.
+As the inspection is done in parallel on each worker node, its overall duration is impacted by
```

Comment:
I can see parallel-over-nodes speeding up aggregation.  I can see serial-over-nodes making it possible for the operator to give up after 8000 pods.  I'm not clear on how you can deliver both fast-parallel aggregation and simultaneously limit to 8000 pods.  Is the operator supposed to check current-pod counts on each node it's considering, and add additional nodes until the current pod-count slightly exceeds 8000, run those nodes in parallel, and then do a second round with additional nodes if pod-completion dropped the gathered data too far below the target 8000?